## Import library

In [1]:
import pandas as pd
import numpy as np

## Import data

In [2]:
df1 = pd.read_csv('conjoint_survey_organic.xlsx - Sheet1.csv') # Hasil survey tanpa ads

In [3]:
df2 = pd.read_csv('conjoint_survey_ads.csv')# Hasil survey dengan ads

## Exploring Data

In [4]:
df1.info() #elihat detail informasi dari hasil survey tanpa ads

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56 entries, 0 to 55
Data columns (total 12 columns):
 #   Column                                                                                                                                                                                                                 Non-Null Count  Dtype 
---  ------                                                                                                                                                                                                                 --------------  ----- 
 0   Timestamp                                                                                                                                                                                                              56 non-null     object
 1   Berapa nomer telepon anda? Nomer ini akan digunakan untuk membagikan GoPay Rp 50.000 per orang, hasil undian untuk 100 orang. Kami hanya akan mengirimkan ke pengis

In [5]:
df2.info() #elihat detail informasi dari hasil survey dengan ads

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 229 entries, 0 to 228
Data columns (total 12 columns):
 #   Column                                                                                                                                                                                                                 Non-Null Count  Dtype 
---  ------                                                                                                                                                                                                                 --------------  ----- 
 0   Timestamp                                                                                                                                                                                                              229 non-null    object
 1   Berapa nomer telepon anda? Nomer ini akan digunakan untuk membagikan GoPay Rp 50.000 per orang, hasil undian untuk 100 orang. Kami hanya akan mengirimkan ke peng

## Explanation

berdasarkan hasil eksplorasi terhadap 2 hasil survey tanpa ads dan dengan ads dapat dilihat banyaknya jumlah data yang diperoleh. Dalam hal in survey berhasil mendapatkan data yang lebih banyak yaitu 229 baris yang mengindikasikan ada 229 user yang berpartisipasi dalam survey, sementara itu survey tanpa ads hanya bisa mengumpulkan 56 data user.

## Concatenate 2 data table

In [6]:
#hilangkan # jiak ingin melihat semua rows data
#pd.set_option('display.max_rows', None)

In [7]:
df =  pd.concat([df1, df2]) # menggabungkan 2 hasil survey
df

Timestamp  \
0    2023-03-08 18:47:23   
1    2023-03-08 18:48:27   
2    2023-03-08 20:17:52   
3    2023-03-13 20:59:49   
4    2023-03-13 21:01:45   
..                   ...   
224  2023-03-18 09:07:04   
225  2023-03-18 11:38:21   
226  2023-03-18 14:45:51   
227  2023-03-19 11:08:15   
228  2023-03-19 15:22:24   

    Berapa nomer telepon anda? Nomer ini akan digunakan untuk membagikan GoPay Rp 50.000 per orang, hasil undian untuk 100 orang. Kami hanya akan mengirimkan ke pengisi kuisioner yang valid, i.e. jawaban tidak random.  \
0                                        08xx336019xxx                                                                                                                                                                      
1                                         08xx10001xxx                                                                                                                                                                      
2                                         08xx56592xxx                                                                                                                                                                      
3                                        08xx391631xxx                                                                                                                                                                      
4                                         08xx21820xxx                                                                                                                                                                      
..                                                 ...                                                                                                                                                                      
224                                       08xx33920xxx                                                                                                                                                                      
225                                       08xx32245xxx                                                                                                                                                                      
226                                       08xx69965xxx                                                                                                                                                                      
227                                       08xx12699xxx                                                                                                                                                                      
228                                       08xx77812xxx                                                                                                                                                                      

    1. Produk manakah yang akan anda beli? (Anda bisa memilih membeli (klik) lebih dari 1 pilihan)  \
0                                                    B                                               
1                                                    C                                               
2                       D. Tidak memilih semua product                                               
3                       D. Tidak memilih semua product                                               
4                       D. Tidak memilih semua product                                               
..                                                 ...                                               
224                     D. Tidak memilih semua product                                               
225                                                  A                                               
226                                                  C                                               
227  

In [8]:
df.info() # informasi semua survey

<class 'pandas.core.frame.DataFrame'>
Index: 285 entries, 0 to 228
Data columns (total 12 columns):
 #   Column                                                                                                                                                                                                                 Non-Null Count  Dtype 
---  ------                                                                                                                                                                                                                 --------------  ----- 
 0   Timestamp                                                                                                                                                                                                              285 non-null    object
 1   Berapa nomer telepon anda? Nomer ini akan digunakan untuk membagikan GoPay Rp 50.000 per orang, hasil undian untuk 100 orang. Kami hanya akan mengirimkan ke pengisi k

# Data processing

## Cleaning Data

In [9]:
# 1. Mengubah nama kolom dari hasil survey yang sudah digabung agar lebih sederhana
new_column = ['Timestamp','User_phone','pertanyaan 1','pertanyaan 2','pertanyaan 3','pertanyaan 4','pertanyaan 5','pertanyaan 6','pertanyaan 7','pertanyaan 8','pertanyaan 9','pertanyaan 10']

In [10]:
df.columns = new_column #mengaplikasikan perubahan nama kolom

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 285 entries, 0 to 228
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Timestamp      285 non-null    object
 1   User_phone     285 non-null    object
 2   pertanyaan 1   285 non-null    object
 3   pertanyaan 2   285 non-null    object
 4   pertanyaan 3   285 non-null    object
 5   pertanyaan 4   285 non-null    object
 6   pertanyaan 5   285 non-null    object
 7   pertanyaan 6   285 non-null    object
 8   pertanyaan 7   285 non-null    object
 9   pertanyaan 8   285 non-null    object
 10  pertanyaan 9   285 non-null    object
 11  pertanyaan 10  285 non-null    object
dtypes: object(12)
memory usage: 28.9+ KB


In [12]:
# Loop melalui kolom pertanyaan 1 - 10 & Hapus baris dengan nilai ', D.' di kolom pertanyaan
for i in range(1, 11):
    kolom = f'pertanyaan {i}'
    df = df[~df[kolom].str.contains(', D.')]

    
# Question, Mengapa jika menggunakan df.drop output yang dihasilkan beda?
# for i in range(1, 11):
#     kolom = f'pertanyaan {i}'
#     df.drop(df[df[kolom].str.contains(', D.')].index, inplace=True)

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 274 entries, 0 to 228
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Timestamp      274 non-null    object
 1   User_phone     274 non-null    object
 2   pertanyaan 1   274 non-null    object
 3   pertanyaan 2   274 non-null    object
 4   pertanyaan 3   274 non-null    object
 5   pertanyaan 4   274 non-null    object
 6   pertanyaan 5   274 non-null    object
 7   pertanyaan 6   274 non-null    object
 8   pertanyaan 7   274 non-null    object
 9   pertanyaan 8   274 non-null    object
 10  pertanyaan 9   274 non-null    object
 11  pertanyaan 10  274 non-null    object
dtypes: object(12)
memory usage: 27.8+ KB


## Explanation
Membersihkan data dari user yang menjawab D bersamaan dengan jawaban lain, user yang memberikan jawaban seperti ini akan dianggap tidak valid dan datanya akan dibuang.

## Create new clean dataframe table

In [14]:
# Membuat dataframe dengan nama kolom sesuai ketentuan
new_df = pd.DataFrame(columns=['user_phone', 'choice', 'skill', 'bentuk_program', 'harga_program'])

In [15]:
# kolom user_phone berdasarkan kolom user_phone pada dataframe sebelumnya
# repeat 30 menunjukan 1 user mewakili 10 data pertanyaan dengan 3 pilihan pada masing2 pertanyaan.
new_df["user_phone"] = df["User_phone"].repeat(30)

In [16]:
new_df.info() #melihat informasi dataframe baru

<class 'pandas.core.frame.DataFrame'>
Index: 8220 entries, 0 to 228
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   user_phone      8220 non-null   object
 1   choice          0 non-null      object
 2   skill           0 non-null      object
 3   bentuk_program  0 non-null      object
 4   harga_program   0 non-null      object
dtypes: object(5)
memory usage: 385.3+ KB


In [17]:
jumlah_unik = new_df['user_phone'].nunique() #mencari jumlah unik dari userphone untuk melihat banyaknya jumlah user

In [18]:
jumlah_unik

274

In [19]:
# Membuat daftar kolom skill untuk 1 user sehingga perlu dikali 274 untuk mendapatkan data dari semua user
new_df["skill"] = ["Create Analytics Dashboard", "Perform Customer Segmentation", "Design AB Test Experimentation",
          "Create Analytics Dashboard", "Design Data Pipeline", "Perform Credit Scoring Analytics",
          "Perform Customer Segmentation", "Perform Customer Segmentation", "Design Data Pipeline",
          "Design AB Test Experimentation","Perform Price Optimization","Perform Credit Scoring Analytics",
         "Design Data Pipeline","Perform Customer lifetime Analysis","Design AB Test Experimentation",
         "Perform Churn Analytics","Perform Customer Segmentation","Create Machine Learning Model",
         "Perform Customer lifetime Analysis","Design Data Pipeline","Deploy Machine Learning Model",
         "Perform Credit Scoring Analytics","Design Data Pipeline","Create Machine Learning Model",
         "Create Analytics Dashboard","Design AB Test Experimentation","Perform Customer lifetime Analysis",
         "Perform Credit Scoring Analytics","Perform Churn Analytics","Perform Churn Analytics"]*274

In [20]:
# Membuat daftar kolom bentuk program untuk 1 user sehingga perlu dikali 274 untuk mendapatkan data dari semua user
new_df["bentuk_program"] =["Tutorial Based","Mentoring Based","Mentoring Based",
                   "Tutorial Based","Mentoring Based","Mentoring Based",
                   "Mentoring Based","Tutorial Based","Mentoring Based",
                   "Mentoring Based","Tutorial Based","Mentoring Based",
                   "Mentoring Based","Tutorial Based","Tutorial Based",
                   "Tutorial Based","Mentoring Based","Mentoring Based",
                   "Tutorial Based","Mentoring Based","Tutorial Based",
                   "Mentoring Based","Mentoring Based","Tutorial Based",
                   "Mentoring Based","Tutorial Based","Mentoring Based",
                   "Mentoring Based","Mentoring Based","Tutorial Based"]*274

In [21]:
# Membuat daftar kolom harga program untuk 1 user sehingga perlu dikali 274 untuk mendapatkan data dari semua user
new_df["harga_program"] = ["Rp 500.000,0","Rp 350.000,0","Rp 300.000,0",
                  "Rp 500.000,0","Rp 300.000,0","Rp 550.000,0",
                  "Rp 350.000,0","Rp 450.000,0","Rp 250.000,0",
                  "Rp 500.000,0","Rp 350.000,0","Rp 350.000,0",
                  "Rp 400.000,0","Rp 300.000,0","Rp 300.000,0",
                  "Rp 450.000,0","Rp 300.000,0","Rp 300.000,0",
                  "Rp 500.000,0","Rp 550.000,0","Rp 350.000,0",
                  "Rp 300.000,0","Rp 550.000,0","Rp 550.000,0",
                  "Rp 250.000,0","Rp 550.000,0","Rp 350.000,0",
                  "Rp 400.000,0","Rp 450.000,0","Rp 500.000,0"]*274

In [22]:
new_df

user_phone choice                               skill  \
0    08xx336019xxx    NaN          Create Analytics Dashboard   
0    08xx336019xxx    NaN       Perform Customer Segmentation   
0    08xx336019xxx    NaN      Design AB Test Experimentation   
0    08xx336019xxx    NaN          Create Analytics Dashboard   
0    08xx336019xxx    NaN                Design Data Pipeline   
..             ...    ...                                 ...   
228   08xx77812xxx    NaN      Design AB Test Experimentation   
228   08xx77812xxx    NaN  Perform Customer lifetime Analysis   
228   08xx77812xxx    NaN    Perform Credit Scoring Analytics   
228   08xx77812xxx    NaN             Perform Churn Analytics   
228   08xx77812xxx    NaN             Perform Churn Analytics   

      bentuk_program harga_program  
0     Tutorial Based  Rp 500.000,0  
0    Mentoring Based  Rp 350.000,0  
0    Mentoring Based  Rp 300.000,0  
0     Tutorial Based  Rp 500.000,0  
0    Mentoring Based  Rp 300.000,0  
..               ...           ...  
228   Tutorial Based  Rp 550.000,0  
228  Mentoring Based  Rp 350.000,0  
228  Mentoring Based  Rp 400.000,0  
228  Mentoring Based  Rp 450.000,0  
228   Tutorial Based  Rp 500.000,0  

[8220 rows x 5 columns]

In [23]:
# membuat list kosong untuk data kolom choice
data_choice = {'choice': []}

In [24]:
# mengambil data untuk list choice 
# Iterasi melalui setiap baris di df
for _, row in df.iterrows():
    user_num = row['User_phone']
    
    # Iterasi melalui setiap kolom pertanyaan di df
    for kolom in df.columns[2:]:
        pertanyaan = row[kolom].split(', ')
        
        # Iterasi melalui setiap pilihan dalam pertanyaan
        for choice in ['A', 'B', 'C']:
            choice_value = 1 if choice in pertanyaan else 0
            
            data_choice['choice'].append(choice_value)

In [25]:
# cek jumlah data choice
len(data_choice['choice'])

8220

In [26]:
# masukan list data choice ke kolom new_df choice
new_df["choice"] = data_choice['choice']

In [27]:
new_df.info() # Melihat informasi dataframe baru

<class 'pandas.core.frame.DataFrame'>
Index: 8220 entries, 0 to 228
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   user_phone      8220 non-null   object
 1   choice          8220 non-null   int64 
 2   skill           8220 non-null   object
 3   bentuk_program  8220 non-null   object
 4   harga_program   8220 non-null   object
dtypes: int64(1), object(4)
memory usage: 385.3+ KB


In [28]:
new_df # cek hasil akhir

user_phone  choice                               skill  \
0    08xx336019xxx       0          Create Analytics Dashboard   
0    08xx336019xxx       1       Perform Customer Segmentation   
0    08xx336019xxx       0      Design AB Test Experimentation   
0    08xx336019xxx       1          Create Analytics Dashboard   
0    08xx336019xxx       0                Design Data Pipeline   
..             ...     ...                                 ...   
228   08xx77812xxx       0      Design AB Test Experimentation   
228   08xx77812xxx       0  Perform Customer lifetime Analysis   
228   08xx77812xxx       1    Perform Credit Scoring Analytics   
228   08xx77812xxx       0             Perform Churn Analytics   
228   08xx77812xxx       0             Perform Churn Analytics   

      bentuk_program harga_program  
0     Tutorial Based  Rp 500.000,0  
0    Mentoring Based  Rp 350.000,0  
0    Mentoring Based  Rp 300.000,0  
0     Tutorial Based  Rp 500.000,0  
0    Mentoring Based  Rp 300.000,0  
..               ...           ...  
228   Tutorial Based  Rp 550.000,0  
228  Mentoring Based  Rp 350.000,0  
228  Mentoring Based  Rp 400.000,0  
228  Mentoring Based  Rp 450.000,0  
228   Tutorial Based  Rp 500.000,0  

[8220 rows x 5 columns]

In [29]:
# Menyimpan DataFrame ke dalam file CSV
#new_df.to_csv('clean_data.csv', index=False)

## Explanation
dataframe baru dibuat dengan nama kolom yang sesuai, dan disesuaikan dengan jumlah user, berdasarkan eksplorasi sebelumnya terdapat 274 user, sehingga perlu dibuat kolom user_phone, skill, bentuk_program, harga_program mewakilai jumah user * 10 pertanyaan * 3 pilihan untuk setiap pertanyaan. Sementara untuk kolom choice menggunakan logika akan bernilai 1 pada pilihan pertanyaan yang dipilih, dengan memperhatikan urutan baris pada kolom pertanyaan. 
jika pertanyaan 1 dataframe 1 = A maka kolom choice dataframe 2 baris 1 = 1,
jika pertanyaan 1 dataframe 1 = B maka kolom choice dataframe 2 baris 2 = 1,
jika pertanyaan 1 dataframe 1 = C maka kolom choice dataframe 2 baris 3 = 1,
jika pertanyaan 2 dataframe 1 = A maka kolom choice dataframe 2 baris 4 = 1,
jika pertanyaan 2 dataframe 1 = B maka kolom choice dataframe 2 baris 5 = 1,
jika pertanyaan 2 dataframe 1 = C maka kolom choice dataframe 2 baris 6 = 1,
jika pertanyaan 3 dataframe 1 = A maka kolom choice dataframe 2 baris 7 = 1,
jika pertanyaan 3 dataframe 1 = B maka kolom choice dataframe 2 baris 8 = 1,
jika pertanyaan 3 dataframe 1 = C maka kolom choice dataframe 2 baris 9 = 1,
dan seterusnya. Selain itu kolom akan bernilai 0